src/main/java/org/apache/commons/compress/archivers/tar/TarArchiveSparseInputStream.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.commons.compress.archivers.tar;
+
+import org.apache.commons.compress.utils.BoundedInputStream;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * For sparse tar entries, there are many "holes"(consisting of all 0) in the file. Only the non-zero data is
+ * stored in tar files, and they are stored separately. The structure of non-zero data is introduced by the
+ * sparse headers using the offset, where a block of non-zero data starts, and numbytes, the length of the
+ * non-zero data block.
+ * This class is used to construct an input stream that combines the "holes" and the non-zero data together using
+ * the sparse headers. When reading from this input stream, the actual data is read out with "holes" and non-zero
+ * data combined together according to the sparse headers.
+ */
```

Comment:
From the description I don't understand why this needs a separate class. Why no have a single `TarArchiveInputStream` that also handles support for sparse entries? Can you elaborate?

Reply:
Of course I can make `TarArchiveInputStream` be able to handle sparse entries, therefore I need to modify the `read()` and `skip()` in `TarArchiveInputStream` when we are reading sparse entries.

The reason why I create a new class `TarArchiveSparseInputStream` is that I don't want to modify `TarArchiveInputStream` too much. I want to try my best to keep the `TarArchiveInputStream` as the original logic as much as possible.

With a separate `TarArchiveSparseInputStream`, I can replace the input stream to a new `TarArchiveSparseInputStream`. I don't need to modify the `read()` or `skip()` in `TarArchiveInputStream`.

Reply:
I'm not sure if creating a single `TarArchiveSparseInputStream` is a good idea or not. Maybe we can discuss more about this.